dapp/request_limitation/count.go

```diff
@@ -1,47 +1,78 @@
 package request_limitation
 
-import "sync"
-
 // a simple increase and decrease
 // request limitation.
 type Count struct {
-	lock              sync.Mutex
-	count             uint
 	max               uint
 	canNotIncreaseErr error
+	increase          chan struct{}
+	decrease          chan struct{}
+	count             chan chan uint
 }
 
 func NewCount(max uint, canNotIncreaseErr error) *Count {
-	return &Count{
-		lock:              sync.Mutex{},
+	c := &Count{
 		max:               max,
-		count:             0,
 		canNotIncreaseErr: canNotIncreaseErr,
+		increase:          make(chan struct{}),
+		decrease:          make(chan struct{}),
+		count:             make(chan chan uint),
 	}
+
+	// state
+	go func() {
+
+		var count uint
+
+		for {
+
+			// exit from go routine
+			if c.increase == nil || c.decrease == nil || c.count == nil {
+				return
+			}
+
+			select {
+			case <-c.increase:
+				count++
+			case <-c.decrease:
+				count--
```

Comment:
uint has a default value of 0

Decreasing it directly can lead to an underflow and `count` would become `4294967295` 

This is likely not an expected behaviour